### ISSUE TITLE
Incorrect pieces data 

### ISSUE THREAD

**Issue by Dim0N22** [elgatito/elementum#65]
WebUI now can visualise pieces information in the torrent panel. Pieces data is fetched using the `/info` handle every 5 seconds.  An example request: 
```
GET http://192.168.1.8:65220/info?torrentid={torrent_hash}&pieces=true&trackers=false
```
There is some bug in the handler which causes inconsistent pieces data returned, for example:
Response 1:
```
    Name:               torrent_name
...
    Pieces:
        ++11++111111111++++11111++1+1++1++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
        ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
        ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```
Response 2 after 5 seconds:
```
    Name:               torrent_name
...
    Pieces:
        ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
        ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
        ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```
Response 3 after 5 more seconds:
```
    Name:               torrent_name

    Pieces:
        +1+1++111111111++++11111++1+1++1++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
        ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
        ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```

You can see that first N pieces are switching between the downloaded / not downloaded states when in reality they all are downloaded.

This causes issues in the WebUI when pieces become missing after showing as downloaded:
![Screenshot 2021-06-07 211420](https://user-images.githubusercontent.com/1761368/121122407-420cdc80-c7d6-11eb-816b-2c1d16246145.png)
![Screenshot 2021-06-07 211451](https://user-images.githubusercontent.com/1761368/121122409-42a57300-c7d6-11eb-87d1-7fa3541c0448.png)
![Screenshot 2021-06-07 211533](https://user-images.githubusercontent.com/1761368/121122410-42a57300-c7d6-11eb-89da-16280e595cc3.png)

I suspect there is some concurrency bug or a similar problem. 

It would be great if you could fix it.
Thanks.

**Comment by antonsoroko** [comment]
is it reproducible bug? e.g. how often it shows wrong pieces info?
also what kind of storage you use? i guess for memory pieces info will be quite "fake".


i did not notice this issue on my hosts. and i do not see any potential issue in the golang code. this info provided by libtorrent.

**Comment by Dim0N22** [comment]
Yes, it is reproducible, but I don't know exact steps to reproduce. It seems to happen to some newly added torrents.
So the workflow is roughly as this:
1. Add a new torrent;
2. Open the pieces panel;
3. Wait for a couple updates. Some of them return incorrect data.

Try adding multiple torrents to check if it happens to you as well.

**Comment by Dim0N22** [comment]
Ok, I think I've found a way of reproducing it. Request pieces data using url
```
http://192.168.1.8:65220/info?torrentid={HASH}&pieces=true&trackers=false
```

Quickly reload the page multiple times, some of the updates will have wrong pieces data.

Here is how I reproduced it after a couple of page reloads:

https://user-images.githubusercontent.com/1761368/122668709-332b1000-d16e-11eb-8467-d8c014ec0c85.mp4

**Comment by antonsoroko** [comment]
with very quick refresh i got
```
    Pieces:
        444444++44++++++4444444444444444++4+4444444+444+444444+44+44444444444444444+++++44444444444444444444
        4444444444444444444444444444+++++++++++++++++++++++++++++++++4444444+++++4444444+44+4444444+44444444
        4444444444444444444444444+++++44++4+++444+++++4+4+4+44++44444+4+4+++++++44444444444444444++4++++++4+
        +4++4+++++4+4+4+44++44444+4+4+++++++444444444444444444++4444444+44++44444444+44+444444444+4+4+++++++
        44444444444444444444+444444+44++44444444+44+444444444+4+4+++++++444444444444444444444444444444444444
        4444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444444
        44444444++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++444+4++4++++++++++++++++++++
        ++++++++++++++++++++++++++++++++++++444++444++++++++++++++++++++++++++++++++++++++++++++++++++
```

so looks like sometimes libtorrent returns wrong data
https://libtorrent.org/reference-Torrent_Status.html#pieces

https://github.com/elgatito/elementum/blob/ce4ef5d84550df4883072095cf60618cb0eeb80f/bittorrent/torrent.go#L2154
https://github.com/elgatito/elementum/blob/ce4ef5d84550df4883072095cf60618cb0eeb80f/bittorrent/torrent.go#L1498
https://github.com/elgatito/elementum/blob/ce4ef5d84550df4883072095cf60618cb0eeb80f/bittorrent/torrent.go#L1482

or it is somehow screwed in cgo interlayer or somewhere else. i honestly don't know.

**Comment by elgatito** [comment]
There is a chance you get pieces information in the middle of some updates with that data.
I would not like to touch the libtorrent logic in Elementum, lt2http code looks more clear for me.

**Comment by antonsoroko** [comment]
Sounds reasonable. If we are going to move to lt2http that does not involve all this swig/cgo magic, some issues will fall away naturally.
